src/nvim/message.c

```diff
@@ -3182,7 +3182,7 @@ bool msg_enable_msgfunc(void)
   // Note: msgfunc is disabled in command line mode
   //       Because default echo is used for command line redraw
   return *p_msgfunc != NUL && !msg_check_loop
-    && !redir_off && !need_wait_return && !starting
+    && !redir_off && !need_wait_return && !starting && !called_emsg
```

Comment:
Hm... I don't think this is the right fix. I think error messages should go to `msgfunc`.

In https://github.com/neovim/neovim/pull/16480#issuecomment-987334632 I called `input("What is your name? ")` which means I should be in command line mode. Then I `set msgfunc=` **while I was still in command line mode**,  but it should not matter in any case since the feature should be completely disabled in command line mode.

The desired outcome for this mode is that no message should be sent, regardless of it being an error or not. It seems like the command line mode detection is incomplete or not enough. Are we considering all the relevant states when `input()`, `getchar()`, etc, are called? Because that should be "command line mode" still.

Reply:
I think error messages should not be passed to `msgfunc`.
Because error messages should not be bypassed.
It confusing users.

> In #16480 (comment) I called input("What is your name? ") which means I should be in command line mode. Then I set msgfunc= while I was still in command line mode, but it should not matter in any case since the feature should be completely disabled in command line mode.

Please read your code more carefully.

```vim
  let name = input("What is your name? ")
  let value = foo()
  set msgfunc=
```

You have called `input()` and `set msgfunc=`, but you are not in command line mode when `set msgfunc=` is executed.
Why?  Because if `input()` is executed and user quit input, it leaves command line mode!

Reply:
Oh you're right, I got confused about command line mode. Sorry.

Reply:
Now I remember why I got confused.

1. I am in command line mode for `input()`
2. `foo()` throws an exception; part of the exception's message is sent to msgfunc, and another part is sent to the output buffer
3. Because the message from `foo()` was sent to the output buffer, I need to press Enter

In step 3 I thought I was still in command line mode due to the need to press Enter, because my cursor did not move back to the window.

`input()` has something to do with this problem because the whole exception is sent to msgfunc if it is in thrown while I am in normal mode. Maybe I have not "exited completely" by the time `foo` is called, as far as the editor is concerned.

Reply:
Press Enter prompt is not command line mode.

Reply:
> Press Enter prompt is not command line mode.

Right! It makes sense now. It just "looks like" it stayed on command line mode because there's no redraw in-between.

The following is with 478db3385

```vim
let g:history = []

fun! Func(...)
  call add(g:history, ["msgfunc", deepcopy(a:000)])
endfun
set msgfunc=Func

autocmd ModeChanged * call add(g:history, ["ModeChanged", deepcopy(v:event)])
fun! OnVimEnter(...)
  autocmd CmdlineLeave * ++once call add(g:history, "CmdlineLeave")
  let name = input("What is your name? ")
  call add(g:history, "Call foo(), mode: " . mode())
  let value = foo()
  set msgfunc=
endfun

autocmd VimEnter * call timer_start(0, function("OnVimEnter"))
```


`[['ModeChanged', {'old_mode': 'n', 'new_mode': 'c'}], 'CmdlineLeave', 'Call foo(), mode: n', ['msgfunc', ['msg_show', 'emsg', [[4, 'Error detected while processing function OnVimEnter:']], v:false]], ['msgfunc', ['msg_clear', 'return_prompt', [], v:false]], ['ModeChanged', {'old_mode': 'c', 'new_mode': 'n'}], ['ModeChanged', {'old_mode': 'n', 'new_mode': 'c'}]]`